---
title: "DE lncRNAs assoicated with RB1 bi-allelic deletions - Figure 5"
output: html_notebook
---

#LOAD LIBRARIES
```{r}
library(rstatix)
library(ggpubr)
library(tidyverse)
library(rsample)  
library(caret)
library(ggpubr)
library(survival)
library(survminer)
library(survMisc)
library(glmnet)
library(vip)
library(ggfortify)
```


#READ IN METADATA FOR RNA-SEQ
```{r}
#Meta data for Deep RNA-seq
anno <- read.table("/Users/ds/Desktop/projects/data/anno/202009_deepRNAseq_sample_full.txt", header =T)
anno <- anno %>% dplyr::select(c(sample_id, wgs_id, biopsy_site, disease_type_SIG_GEX, ID_patient, RB1,tumor_purity_rna)) 
cf <- read.table("/Users/ds/Desktop/projects/mCRPC_lncRNA_analysis/data/CIBERSORTx_Job33_Adjusted.txt", sep = "\t", header = T)
cf$sample_id <- gsub("\\.", "-", cf$Mixture)
row.names(cf) <- cf$Mixture
cf <- inner_join(x = cf, y = anno, by  = c("sample_id" = "sample_id"))
row.names(cf) <- cf$sample_id
#Omit samples without WGS data
cf <- na.omit(cf)
#Format samples with RB1 bi-allelic deletions v.s monoallelic/WT
cf$RB1 <- ifelse(cf$RB1 == 0, "Del", "WT.Mono")

```

#EDGER DIFFERENTIAL GENE EXPRESSION ANALYSIS OF LNCRNAS AND RB1 STATUS
```{r}
#Read in raw counts, convert to matrix with rows as gene symbols, columns as samples
counts <- read.table("/Users/ds/Desktop/projects/data/rna/mCRPC_RNA_counts_genename.txt", header = T)
counts <- counts[!duplicated(counts$Genename),]
row.names(counts) <- counts$Genename
counts <- counts[,-1] %>% as.matrix()
colnames(counts) <- gsub("\\.", "-", colnames(counts))  

#Keep matching sample ids of RNA-seq and WGS for cmp annotation
ids <- intersect(colnames(counts),row.names(cf))
counts <- counts[,ids]
cf <- cf[ids,]

#Create dge object - create counts matrix and sample anno df with library sizes
x <- DGEList(counts = counts)

#Add cluster variable to designate group membership to sample df
x$samples <- cbind(x$samples, dplyr::select(cf, c( biopsy_site, disease_type_SIG_GEX, RB1,tumor_purity_rna )))

#Remove genes with low expression across samples
keep.exprs <- filterByExpr(x, min.count = 5)
x <- x[keep.exprs,, keep.lib.size = F]

#Calculate effective library size using EdgeR TMM method
x <- calcNormFactors(x, method = "TMM")

#Create design matrix (w/o intercept) specifying group membership - tumor purity as covariate
design <- model.matrix(~ 0 + RB1 + tumor_purity_rna, data = cf) 

colnames(design) <- c("Del", "WT.Mono", "Purity") 

# #Create contrasts to calculate
contr.matrix <- makeContrasts(Del - WT.Mono, levels = colnames(design))

#Normalize counts to log CPM using voom
v <- voom(x, design = design)

#Fit LM according to design
vfit <- lmFit(v, design)

#Calculate contrasts i.e logFC for genes across groups
vfit <- contrasts.fit(vfit, contrasts=contr.matrix)

#Calculate t-stats and p vals using Emp bayes method
efit <- eBayes(vfit)


#Get list of all genes, arrange in decreasing order - FDR 10% , abs logFC > 1, merge with lncRNA markers
bulk_deg <- topTable(efit, number = Inf, coef = 1) %>% data.frame(Genes = row.names(.))  %>% arrange(-logFC)
bulk_deg <- bulk_deg %>% dplyr::filter(adj.P.Val < 0.1 & abs(logFC) > 1) %>% left_join( y = final_markers, by = c("Genes" = "gene"))
# bulk_deg <- bulk_deg %>% dplyr::filter(adj.P.Val < 0.1 & abs(logFC) > 1) %>% left_join( y = pcg_markers, by = c("Genes" = "gene"))

#Format deg list
bulk_deg$cluster <- bulk_deg$cluster %>% as.character()
bulk_deg <- bulk_deg %>% mutate(Cluster = ifelse(is.na(cluster), "Unknown", cluster)) %>% mutate(Alpha = ifelse(Cluster == "Unknown", 0.8, 1))



```


#DEG PLOT
```{r}
#Plotting
plot <- ggplot(bulk_deg, aes(x = logFC, y = -log10(adj.P.Val), color = Cluster, alpha = Alpha, label = Genes) ) + geom_point() + scale_color_manual(values = c("purple", "orange", "red", "pink", "darkgreen","navy" , "grey")) + geom_hline(yintercept = -log10(0.1), linetype= 'dashed') + geom_vline(xintercept = c(-1,1), linetype = 'dashed') + geom_text_repel(data = dplyr::filter(bulk_deg, abs(logFC) > 1 &adj.P.Val < 0.1 & Cluster != "Unknown"), force = 0.8, size = 2.5) + ggtitle(label = "Protein coding genes associated with bi-allelic RB1 loss") + theme_classic()
plot
ggsave(plot = plot, filename = "/Users/ds/Desktop/plot.pdf", width = 10)

#Write to disk
write.csv(x = bulk_deg, file = "/Users/ds/Desktop/Supplementary.Table5.csv",quote = F, row.names = F)

```


```{r}
mcrpc_rna <- read.table(file = "/Users/ds/Desktop/projects/data/rna/mCRPC_RNA_counts_genename.txt", header = T)
mcrpc_rna <- mcrpc_rna[!duplicated(mcrpc_rna$Genename),]




row.names(mcrpc_rna) <- mcrpc_rna$Genename
mcrpc_rna <- mcrpc_rna[,-1]
keep.exprs <- filterByExpr(as.matrix(mcrpc_rna),min.count = 5 )




mcrpc_rna <- mcrpc_rna[keep.exprs,]
mcrpc_rna <- edgeR::cpm(mcrpc_rna, log = T) %>% t() %>% data.frame()#Log normalize for heatmap
#mcrpc_rna <- mcrpc_rna %>% t() %>% data.frame()#Counts for DEG analysis
row.names(mcrpc_rna) <- gsub("\\.", "-",  x = row.names(mcrpc_rna))
#mcrpc_rna <- mcrpc_rna %>% mutate(cases = row.names(.), batch = "mCRPC")



anno <- read.table("/Users/ds/Desktop/projects/data/anno/202009_deepRNAseq_sample_full.txt", header =T)



anno <- anno %>% dplyr::select(c(sample_id, wgs_id, biopsy_site, disease_type_SIG_GEX, ID_patient,tumor_purity_rna, RB1 ))

clin <- read_csv(file = "/Users/ds/Desktop/projects/data/anno/Chen_2019_European_Eurology_Supp4.csv", col_names = T)
clin$Enzalutamide.resistant  <- clin$`Enzalutamide resistant`
clin <- na.omit(clin)
```

```{r}
rbs <- read_csv(file = "/Users/ds/Desktop/Supplement/chen.rb1.loss.csv", col_names = T)

sig.up <- rbs$`Highly-expressed in RB1-loss:` %>% na.omit()
sig.down <- rbs$`Low-expressed in RB1-loss:` %>% na.omit()


```

```{r}
tme.up <- bulk_deg %>% dplyr::filter(logFC > 0 & (!Cluster %in% c("tumor", "Unknown", "erythroid")) ) %>% .$Genes

up <- bulk_deg %>% dplyr::filter(logFC > 0 & Genes %in% linc$gene.name) %>% .$Genes
#up <- bulk_deg %>% dplyr::filter(logFC > 0 & Genes %in% pcg$gene.name) %>% .$Genes


tumor.down <- bulk_deg %>% dplyr::filter(logFC < 0 & (Cluster %in% c("tumor")) ) %>% .$Genes


```


```{r}

gsva.es <- gsva(t(mcrpc_rna), gset.idx.list = list(tme.up), method = 'zscore')

df <- t(gsva.es) %>% data.frame(sample_id = row.names(.), TME.pcg = .)

gsva.es <- gsva(t(mcrpc_rna), gset.idx.list = list(c(sig.up)), method = 'zscore')

df <- t(gsva.es) %>% data.frame(sample_id = row.names(.), Pca.pcg = .) %>% merge(df) %>% merge(anno) %>% inner_join(y = clin, by = c("ID_patient" = "Patient.ID"))



df$RB1 <- as.factor(ifelse(df$RB1 == 0, "Deletion", "WT/Mono"))
df$Enzalutamide.resistant <- ifelse(df$Enzalutamide.resistant == TRUE, "Enzalutamide.resistant", "Enzalutamide.naive")

df <- df %>% mutate(Group = as.factor(ntile(TME.pcg, 2)))

df
p1 <- coxph(Surv(time =OS.mCRPC , event = Event)~ Group, data = df) %>% ggforest(fontsize = 1)
p2 <- coxph(Surv(time =OS.mCRPC , event = Event)~ Pca.pcg, data = df) %>% ggforest(fontsize = 1)
p1
p3 <-coxph(Surv(time =OS.mCRPC , event = Event)~ Group  +ECOG.PS + PSA.At.Biopsy+ LDH.at.biopsy + ALP + Hemoglobin.at.biopsy + Disease.Site + Enzalutamide.resistant  , data = df) %>% ggforest(fontsize = 1) 
p4 <- coxph(Surv(time =OS.mCRPC , event = Event)~ Pca.pcg  +ECOG.PS + PSA.At.Biopsy+ LDH.at.biopsy + ALP + Hemoglobin.at.biopsy + Disease.Site + Enzalutamide.resistant  , data = df) %>% ggforest(fontsize = 1) 
p3

pdf("/Users/ds/Desktop/plot.pdf", width = 10, height = 10)
grid.arrange(p1, p3, ncol = 1)
dev.off()


tme.up

```



```{r}
celltypes <- c("CD4" , "CD8" , "CTL1" , "CTL2" , "ImmatureB" , "Mature" , "mDC" , "memBcell" , "Mono1" , "Mono2" , "Mono3"  , "Monocyte" , "NK" , "TAM" , "TIM" , "TregR" , "TregA" , "NKT" , "ProB" , "Th117" , "pDC" , "Progenitors",  "Osteoblasts" , "Osteoclasts" ,"Pericytes" , "Endothelial" , "Erythroid" )



#Loop across all celltypes in cibersort results df
res_cf <- do.call(rbind, lapply(c(celltypes), FUN = function(m){
  
  
  response <- cf[,m]
  
  #Fit linear regression model with covariate as bone met site 
  model <- lm(formula = response ~ cf$RB1 + cf$tumor_purity_rna) 
  #Get coeff for Bone and pvalue
  coef <- summary(model)$coefficients[, "Estimate"][2]
  pval <- summary(model)$coefficients[, "Pr(>|t|)"][2]
  
  return(  data.frame(response, rep(m, length(response)), rep(coef, length(response)) , rep(pval, length(response)), cf$Mixture, cf$RB1) )

 })) %>% data.frame()


colnames(res_cf) <- c("Fraction", "Cell", "Coefficient", "P.value", "Mixture", "RB1")

#FDR correction across celltypes
res_cf$Adj.Pvalue <- rep(p.adjust(unique(res_cf$P.value), method = "BH"), each = nrow(cf))
res_cf$Sig <-  ifelse(res_cf$Adj.Pvalue < 0.1, "*", "ns")

res_cf %>% dplyr::filter(Sig == "*")

res_cf %>% dplyr::filter(Sig == "*") %>% dplyr::filter(Cell %in% celltypes) %>% ggplot( aes(y = Cell, x = Fraction * 100, fill = RB1, alpha = Sig)) + geom_boxplot(outlier.colour = NA ) +  scale_fill_manual(values= c("darkred", "darkblue")) + ylab("Cells") + xlab("Percentage") + scale_alpha_manual(values = c(1,0.1)) + ggtitle("Comparison of cell fractions in mCRPC RNA-seq")

```



```{r}

a <- gsva(t(mcrpc_rna), gset.idx.list = list(tumor.down), method = 'zscore') %>% t()
b <- gsva(t(mcrpc_rna), gset.idx.list = list(sig.up), method = 'zscore') %>% t()
c <- gsva(t(mcrpc_rna), gset.idx.list = list(dplyr::filter(gs, gs_name == "HALLMARK_TNFA_SIGNALING_VIA_NFKB")$gene_symbol), method = 'zscore') %>% t()




p <- data.frame(RBS.high.lncRNA = a,RBS.upregulated.genes = b,Hallmark.TNFA.Signaling.Via.NFKB = c) %>% gather(key = "Signature", value = "GEX", -RBS.high.lncRNA) %>% ggplot(aes(x = RBS.high.lncRNA
, y = GEX, color = Signature)) %>% facet(facet.by = "Signature") + geom_point() + geom_smooth(method = 'lm') + stat_cor() + scale_color_manual(values = c("darkred", "darkblue")) + theme_classic() + ylab("Expression") + xlab("RB1 loss upregulated TME-lncRNAs")

p



```



```{r}

Idents(bmet) <- "biosample_id"

mat =  AverageExpression(object = bmet)



res <- do.call(rbind, lapply(pathways, FUN = function(p){
  
  print(p)
  g <- gs %>% dplyr::filter(gs_name == p) %>% .$gene_symbol
  df <- mat$RNA %>% gsva( gset.idx.list = list(A = tme.up, B = up, C = g), method = 'zscore') %>% t() %>% data.frame()
  df <- data.frame(Pathway = p, PCC.up = cor.test(df$A,df$C)$estimate, Pval.up = cor.test(df$A,df$C)$p.value, PCC.down = cor.test(df$B,df$C)$estimate, Pval.down = cor.test(df$B, df$C)$p.value)
  return(df)
}))


res %>% arrange(-PCC.up)
res.up <- arrange(res, -PCC.up)[1:10,c("Pathway", "PCC.up")]
colnames(res.up) <- c("Pathway", "PCC")

res.down <- arrange(res, -PCC.down)[1:10,c("Pathway", "PCC.down")]
colnames(res.down) <- c("Pathway", "PCC")

p <- rbind(res.down, mutate(res.up, PCC = PCC * -1)) %>% mutate(Group = ifelse(PCC > 0, "All RB1-loss upregulated protein coding genes", "RB1-loss upregulated TME-protein coding genes")) %>% mutate(Pathway = paste0(Pathway,"_", c(1:20))) %>% ggplot(aes(y = reorder(Pathway, PCC), x = PCC, fill = Group)) + geom_col() + xlim(c(-1,1)) + scale_x_continuous(breaks = seq(-1,1,0.1),labels = seq(-1,1,0.1) %>% abs()) + ylab("Pathway") + scale_fill_manual(values = c("darkred", "navy")) + theme_classic() 

p

ggsave(filename = "/Users/ds/Desktop/plot.pdf", plot = p, width = 10)


```


```{r}

up
query <- linc[linc$gene.name %in% up,'gene', drop = T] 

query <- EnsDb.Hsapiens.v86 %>% promoters() %>% data.frame() %>% dplyr::filter(gene_id %in% query) %>% GRanges()

seqlevelsStyle(query) <- "UCSC"



seq <- getSeq(BSgenome.Hsapiens.UCSC.hg38, query)
names(seq) <- query$tx_id
writeXStringSet(seq, filepath = "/Users/ds/Desktop/rb1.up.fa")


query <- linc[linc$gene.name %in% tme.up,'gene', drop = T] 

query <- EnsDb.Hsapiens.v86 %>% promoters() %>% data.frame() %>% dplyr::filter(gene_id %in% query) %>% GRanges()

seqlevelsStyle(query) <- "UCSC"



seq <- getSeq(BSgenome.Hsapiens.UCSC.hg38, query)
names(seq) <- query$tx_id
writeXStringSet(seq, filepath = "/Users/ds/Desktop/rb1.tme.up.fa")




```


```{r}
motif <- read_tsv(file = "/Users/ds/Desktop/RB1.all.up.tsv", col_names = T)



motif.up <- motif %>% mutate(Sig = -`log_adj_p-value`) %>% arrange(-Sig) %>% mutate(TF = gsub(pattern = "_HUMAN.*", replacement = "", x = motif_id)) %>% dplyr::filter(db_index == "1") %>% head(n=20) 
motif.up

motif <- read_tsv(file = "/Users/ds/Desktop/RB1.tme.up.tsv", col_names = T)

motif.tme.up <- motif %>% mutate(Sig = -`log_adj_p-value`) %>% arrange(-Sig) %>% mutate(TF = gsub(pattern = "_HUMAN.*", replacement = "", x = motif_id)) %>% dplyr::filter(db_index == "1") %>% head(n=20) 

p <- rbind(motif.up, mutate(motif.tme.up, Sig = Sig * -1)) %>% mutate(Group = ifelse(Sig > 0, "All RB1-loss upregulated lncRNAs", "RB1-loss upregulated TME-lncRNAs")) %>% ggplot(aes(y = reorder(TF, Sig), x = Sig, fill = Group)) + geom_col() + xlim(c(-100,100)) + scale_x_continuous(breaks = seq(-100,100,10),labels = seq(-100,100,10) %>% abs()) + ylab("TF Motif") + scale_fill_manual(values = c("darkred", "navy")) + theme_classic() + xlab("-log10(Adj Pvalue)")

p

rbind(motif.up, mutate(motif.tme.up, Sig = Sig * -1)) %>% mutate(Group = ifelse(Sig > 0, "All RB1-loss upregulated lncRNAs", "Immune RB1-loss upregulated lncRNAs")) 
motif.tme.up



```



```{r}
clin <- read_tsv(file = "/Users/ds/Desktop/prad_su2c_2019_clinical_data.tsv", col_names = T)
clin <- clin[!is.na(clin$`Overall Survival (Months)`),]

rna <- read_table(file = "/Users/ds/Desktop/data_mrna_seq_fpkm_polya.txt", col_names = T)
rb1 <- read_table(file = "/Users/ds/Desktop/data_cna.txt", col_names = T)

clin <- clin %>% mutate(Sample.ID =gsub(pattern = "-|_", replacement = "\\.", x =`Sample ID`  ))

colnames(rna) <- gsub(pattern = "-|_", replacement = "\\.", x =colnames(rna)  )
colnames(rb1) <- gsub(pattern = "-|_", replacement = "\\.", x =colnames(rb1)  )

rb1 <- rb1 %>% dplyr::filter(Hugo.Symbol == "RB1") %>% t() %>% data.frame(RB1 = .) %>% dplyr::filter(RB1 == -2) %>% row.names()

clin <- clin %>% dplyr::filter(Sample.ID %in% colnames(rna))

clin <- clin %>% mutate(RB1 = ifelse(Sample.ID %in% rb1, 0,1)) 
rna <- rna[!duplicated(rna$Hugo.Symbol),] %>% data.frame()
row.names(rna) <- rna$Hugo.Symbol
rna <- rna[,-1]

rna <- log1p(rna)


```


```{r}


clin <- read_csv(file = "/Users/ds/Desktop/projects/data/anno/Nat_Gen_sampleanno.csv", col_names = T)

rna <- read_table(file = "/Users/ds/Desktop/projects/data/rna/2018_04_15_matrix_rna_counts.txt")

rna

rna <- rna[!duplicated(rna$symbol),] %>% data.frame()



rna
row.names(rna) <- rna$symbol
rna <- rna[,-1]
keep.exprs <- filterByExpr(as.matrix(rna),min.count = 5 )
rna



rna <- rna[keep.exprs,]

rna
rna <- edgeR::cpm(rna, log = T) %>% t() %>% data.frame()#Log normalize for heatmap
#mcrpc_rna <- mcrpc_rna %>% t() %>% data.frame()#Counts for DEG analysis
row.names(rna) <- gsub("\\.", "-",  x = row.names(rna))
#mcrpc_rna <- mcrpc_rna %>% mutate(cases = row.names(.), batch = "mCRPC")


clin

```


```{r}

pathways

df <- gsva(as.matrix(t(rna)), gset.idx.list = list(A = dplyr::filter(gs, gs_name == "HALLMARK_IL6_JAK_STAT3_SIGNALING")$gene_symbol, B = sig.up), method = 'zscore') %>% t()


#df <- gsva(as.matrix(rna), gset.idx.list = list(c("CD3D", "CTLA4")), method = 'zscore') %>% t()


#df <- df %>% data.frame( Sample.ID = row.names(.))%>% merge(clin, by = c("Sample.ID" = "Sample.ID"))

df <- df %>% data.frame( Sample = row.names(.))%>% merge(clin, by = c("Sample.ID" = "Sample"))

df %>% ggplot(aes(x = A, y = B)) +geom_point() + geom_smooth(method = "lm") + stat_cor()


df$Status <- ifelse(df$`Overall Survival Status` == "1:DECEASED", 1, 0) 
#coxph(Surv(time =`Overall Survival (Months)` , event = Status)~ TME.lncRNA, data = df) %>% ggforest(fontsize = 1)

df
df %>% ggplot(aes(y = A, x = as.factor(`RB1 inactivation`))) + geom_boxplot() + geom_jitter(width = 0.3)

t.test( df$A~ as.factor(df$`RB1 inactivation`))


res <- do.call(rbind, lapply(pathways, FUN = function(p){
  
  print(p)
  g <- gs %>% dplyr::filter(gs_name == p) %>% .$gene_symbol
  df <-  gsva(expr = as.matrix(t(rna)) ,gset.idx.list = list(A = g, B = sig.up ), method = 'zscore') %>% t() %>% data.frame()
  df <- data.frame(Pathway = p, PCC.up = cor.test(df$A,df$B)$estimate, Pval.up = cor.test(df$A,df$B)$p.value)
  return(df)
}))

res %>% arrange(-PCC.up)
```

