---
title: "R Notebook"
output: html_notebook
---

```{r}


library(Gviz)
library(GenomicRanges)
library(TxDb.Hsapiens.UCSC.hg38.knownGene)
library(Signac)
library(tidyverse)
library(EnsDb.Hsapiens.v86)
```



```{r}
GOI <- "COLCA1"
chr <- "chr11"
Geneid <- 399948 
target <- EnsDb.Hsapiens.v86 %>% ensembldb::filter(~ symbol == GOI) %>% genes()

#target <- target %>% resize(width = width(target)+ 10000,fix = "center" )

#target <- target[1]
#target <- GRanges(seqnames = '22', ranges = IRanges(start = 23300000, end = 23800000))
```

```{r}
dss <- read_csv(file = "/Users/ds/Desktop/projects/lncRNA/results/DSS_mCRPC_BN.csv", col_names = c("chrom", "start", "end", "Diff", "Stat"))
dss$chrom <- paste0("chr", dss$chrom)
dss <- dss %>% dplyr::filter(chrom == chr) %>% dplyr::select(-Stat) %>% mutate(Diff = Diff * 100) %>% GRanges()


dTrack <- DataTrack(dss, name = "Methylation", background.title = "DarkOrchid4", fill = "DarkOrchid4")

plotTracks(list(dTrack), from = start(target), to = end(target), type = c("histogram"), ylim = c(-30,0))



```




```{r}
txdb <- TxDb.Hsapiens.UCSC.hg38.knownGene %>% exons(columns=c("GENEID", "EXONID", "TXNAME" ))



txdb <- S4Vectors::expand(txdb, c("GENEID", "EXONID", "TXNAME"))
```


```{r}
gene <- txdb[txdb$GENEID == Geneid] %>% data.frame()

colnames(gene) <- c("chromosome", "start", "end", "width", "strand", "gene", "exon", "transcript")



hmr <- read_tsv(file = "/Users/ds/Desktop/projects/lncRNA/data/hmr.bed", col_names  = c("chrom", "start", "end"))
hmr <- hmr %>% dplyr::filter(chrom == chr) %>% dplyr::select(c(chrom, start, end)) %>% GRanges()



metarbs <- read_tsv(file = "/Users/ds/Desktop/projects/lncRNA/data/AR.chip.hg38.bed", col_names = c("chrom", "start", "end"))
metarbs <- metarbs %>% dplyr::filter(chrom == chr) %>% dplyr::select(c(chrom, start, end)) %>% GRanges()

metfrbs <- read_tsv(file = "/Users/ds/Desktop/projects/lncRNA/data/FOXA1.chip.hg38.bed", col_names = c("chrom", "start", "end"))
metfrbs <- metfrbs %>% dplyr::filter(chrom == chr) %>% dplyr::select(c(chrom, start, end)) %>% GRanges()

h3ac <- read_tsv(file = "/Users/ds/Desktop/projects/lncRNA/data/H3K27ac.chip.hg38.bed", col_names = c("chr", "start", "end"))
h3ac <- h3ac %>% dplyr::select(c(chr, start, end)) %>% GRanges()
h3ac <- subsetByOverlaps(x = h3ac, ranges = GRanges(dplyr::filter(linc, gene.name == GOI)))
```


```{r}
#Plot
grtrack <- GeneRegionTrack( genome = "hg38",range = gene, chromosome = chr , name = GOI, transcriptAnnotation = "transcript", background.panel = "lightyellow", background.title = "darkblue", fill = "darkblue", col = "darkblue")

atrack <- AnnotationTrack(hmr, name = "HMR", background.title = "darkred",  background.panel ="lightgrey", fill = "darkred", col = "darkred")

arbtrack <- AnnotationTrack(metarbs, name = "AR", background.title = "darkgreen" , background.panel = "lightgrey", fill = "darkgreen", col = "darkgreen")


frbtrack <- AnnotationTrack(metfrbs, name = "FOXA1", background.title = "darkorange" , background.panel = "lightgrey", fill = "darkorange", col = "darkorange")

atactrack <- AnnotationTrack(h3ac, name = "K27ac", background.title = "black" , background.panel = "lightgrey", fill = "black", col = "black", )




gtrack <- GenomeAxisTrack()
itrack <- IdeogramTrack(genome = "hg38", chromosome = chr)


```


```{r}
pdf("/Users/ds/Desktop/plot.pdf", width = 10)
plotTracks(list(itrack,gtrack, dTrack,arbtrack,frbtrack, atactrack, atrack,grtrack), from = start(target) -10000, to = end(target) + 10000, type = "histogram",ylim = c(-30,0))
dev.off()




```

