---
title: "Create consensus set of RNA, deep RNA, and DNAm HMR"
output: html_notebook
author: Debanjan Saha
---
#Load libraries
```{r}
library(tidyverse)
library(ggpubr)
library(edgeR)
library(ggrepel)
library(viridis)
library(DiffBind)


```

#Create deep RNA CPM txt file from raw counts
```{r}
#Processing deep RNA-seq count data and annotations
counts <- read.table("/Users/ds/Desktop/projects/data/rna/mCRPC_RNA_counts_genename.txt", header = T)
clin <- read.table("/Users/ds/Desktop/projects/data/anno/202009_deepRNAseq_sample_full.txt", header = T) %>% dplyr::select(c(sample_id, wgs_id, MLH1)) 



#Remove duplicates and convert to matrix

counts <- counts[!duplicated(counts$Genename),]
row.names(counts) <- counts$Genename
counts <- as.matrix(counts[,-1])

# counts <- as.matrix(counts)
# 
# #Match RNA ID with WGS ID and keep only matched samples and remove duplicates - 61 samples remaining
# keep <- data.frame(sample_id = gsub(pattern = "\\.", replacement = "-", x =colnames(counts) )) %>% merge(clin) %>% mutate(Index = 1:74) %>% na.omit()
# keep <- keep[!duplicated(keep$wgs_id),]
# 
# #Rename columns to WGS ID
# counts <- counts[,keep$Index]
# colnames(counts) <- keep$wgs_id



#Filter genes with lowly expressed counts and CPM normalize - ~30K genes remaining
keep <- filterByExpr(counts)
rna_cpm <- cpm(y = as.matrix(counts[keep,])) %>% data.frame(Gene = row.names(.)) %>% dplyr::select(Gene, everything())


rna_cpm
#Save to disk
write.table(rna_cpm, file= "/Users/ds/Desktop/projects/data/rna/bulkRNA_mCRPC_CPM_gene.txt", quote = F, sep = "\t", row.names = F, col.names = T)


clin
```


#Create regular RNA CPM txt file from raw counts
```{r}
#Processing deep RNA-seq count data and annotations
counts <- read.table("/Users/ds/Desktop/projects/data/rna/2018_04_15_matrix_rna_counts.txt", header = T)
clin <- read.csv("/Users/ds/Desktop/projects/data/anno/Nat_Gen_supp_T1.csv", header = T) %>% dplyr::select(IDENTIFIER)



#Convert to matrix
row.names(counts) <- counts$symbol
counts <- as.matrix(counts[,-1])
counts <- as.matrix(counts)

#Convert colnames
colnames(counts) <- gsub(pattern = "\\.", replacement = "-", x =colnames(counts))



#CPM normalize 
#keep <- filterByExpr(counts)
rna <- cpm(y = as.matrix(counts[,])) %>% data.frame(Gene = row.names(.)) %>% dplyr::select(Gene, everything())


#Save to disk
#write.table(rna[,1:50], file= "/Users/ds/Desktop/projects/data/rna/2018_04_15_matrix_rna_CPM_genename_1:50.txt", quote = F, sep = "\t", row.names = F, col.names = T)

#write.table(rna[,c(1, 51:100)], file= "/Users/ds/Desktop/projects/data/rna/2018_04_15_matrix_rna_CPM_genename_51:99.txt", quote = F, sep = "\t", row.names = F, col.names = T)

write.table(rna, file= "/Users/ds/Desktop/projects/data/rna/2018_04_15_matrix_rna_CPM_genename.txt", quote = F, sep = "\t", row.names = F, col.names = T)


```



#Create DNAm HMR txt object
```{r}
#Create GRanges object of rHMR for Aggregate_cpg.R script

#Read in rHMR data from Nature Genetics WGBS Supp paper
dnam <- read.csv("/Users/ds/Desktop/projects/data/bs/rHMR_mCRPC.csv", header = T, skip = 1)



#Fix column ordering to bed file
dnam <- apply(dnam, MARGIN = 1,FUN = function(row){
  
  if(row[2] > row[3]){
    return(row[c(1,3,2)])
  } else{
      return(row[c(1,2,3)])
    }
  
  }) %>% t() %>% data.frame() 



#Convert coordiantes to numerics
dnam$X2 <- as.numeric(dnam$X2)
dnam$X3 <- as.numeric(dnam$X3)

dnam
options(scipen = 999)#Avoid scientific notation

#Sort and write to disk
dnam <- arrange(dnam, X1, X2)
colnames(dnam) <- c("seqnames", "start", "end")

dnam <- dnam %>% mutate(Coord = paste0(seqnames, ".", start, ".", end))

dnam <- dnam[!duplicated(dnam$Coord),c("seqnames", "start", "end")]


#%>% saveRDS("/Users/ds/Desktop/scratch/SV_association/HMR/rHMR.rds")#Use methylkit, aggregate.R to find methylation

write.table(dnam, file = "/Users/ds/Desktop/DNAm_datasets/HMR.bed", quote = F, sep = "\t", row.names = F, col.names = F)


#Subset DNAm samples to those with matched WGS (n= 64)
wgs_id <- clin %>% select(c(wgs_id, wgbs)) %>% filter(wgbs == TRUE) %>% .$wgs_id

dnam <- read.table("/Users/ds/Desktop/scratch/SV_association/HMR/all.rHMR.txt", header = T)
row.names(dnam) <- dnam$Coord
dnam <- dnam[,-1]
colnames(dnam) <- colnames(dnam) %>% gsub(pattern = "\\.", replacement = "-")


write.table(dnam[,wgs_id], file = "/Users/ds/Desktop/scratch/SV_association/HMR/rHMR_wgs_match.txt", quote = F, sep = "\t", row.names = T, col.names = T)

```


#Create ATACseq txt file
```{r}
#Load in binding affinity matrix from Diffbind output
dba_res <- readRDS(file = "/Users/ds/Desktop/ATAC_mCRPC_DBA_results.rds") 


#Extract Consensus peakset (minoverlap = 0.5) - RPKM if ATAC - Beta if DNAm
cons <- dba.peakset(DBA = dba_res, peaks = NULL, bRetrieve=T) %>% data.frame()

atac_rpkm <- cons %>% mutate(Coord = paste0(seqnames, ".", start, ".", end)) %>% dplyr::select(c(Coord, starts_with(c("DTB", "PR"))))


final_markers %>% dplyr::filter(cluster != 'tumor')

```


```{r}
#Format to match wgs samples - for SV association analysis
txt <- cons %>% dplyr::select(starts_with(c("DTB", "PR")))
row.names(txt) <- cons$Coord 
t <- data.frame(sample_id = colnames(txt) %>% gsub(pattern = "\\.", replacement = "-") %>% gsub(pattern ="PRO-", replacement = "Pro"), index = 1:length(colnames(txt))) %>% merge(clin)
txt <- txt[,t$index]
colnames(txt) <- t$wgs_id

#Subset common samples w/ RNA and ATAC
rna_cpm <- read.table(file= "/Users/ds/Desktop/scratch/txt/mCRPC_CPM_filter.txt", header = T, )
colnames(rna_cpm) <- gsub("\\.", "-", colnames(rna_cpm)) 

common <- intersect(colnames(txt), colnames(rna_cpm))


write.table(rna_cpm[,common], file= "/Users/ds/Desktop/scratch/txt/mCRPC_CPM_filter_matchATAC.txt", quote = F, sep = "\t", row.names = T, col.names = T)

txt[,common] %>% write.table("/Users/ds/Desktop/scratch/txt/ATAC_mCRPC_consensus.txt", quote = F, sep = "\t", row.names = T, col.names = T)

```



